Track user clicks on Links/Buttons in disconnect modal footer

jim-coffey opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* user should see no difference
* when clicking any link or button in the footer of the disconnect dialog - an event should be recorded so we can see how users are interacting with the modal options (see test plan for details)

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
We are tracking some more clicks but no major change to how or what we track.

#### Testing instructions:

You can see if tracking events are being triggered by looking for the tracking pixels in the Network and checking query parameters - filter network traffic by `t.gif`. Or enabling logging in the browser console, as detailed here PCYsg-nSS-p2. Or visit the internal live tracker website.

* with Jetpack development running locally
* open /wp-admin
* navigate to -> Jetpack -> Dashboard
* click 'Manage Site Connection'
* click the `Jetpack connection` link and check you see the `jetpack_disconnect_dialog_click_learn_about` event
* click the `contact Jetpack support` link and check you see the `jetpack_disconnect_dialog_click_support` event
* click the `Stay Connected` button and check the modal still closes and you see the `jetpack_disconnect_dialog_click_stay_connected` event
* reopen the modal, and click the `Disconnect` button and check you still disconnect and see the `jetpack_disconnect_dialog_click_disconnect` event

Repeat the above from the -> Plugins : Jetpack : `Disconnect and Deactivate` - the events are the same but the `context` of the tracking should be `plugins` rather than `jetpack`

For reference, the links and buttons are illustrated here:
![disconnect_modal](https://user-images.githubusercontent.com/8742105/151022993-540f369f-d276-4aeb-bba3-44794fd68f4e.png)

#### Unit Tests

Unit Tests have been updated and can be checked by running `pnpm t` from the `projects/js-packages/connection/` folder.

jim-coffey committed:
[not verified] Track user clicks on disconnect dialog modal

jim-coffey committed:
Fix and add some unit tests

jim-coffey committed:
Changelog

jboland88 commented on projects/js-packages/connection/components/disconnect-dialog/test/step-disconnect.jsx:
```diff
@@ -47,13 +52,37 @@ describe( 'StepDisconnect', () => {
+		const links = wrapper.find( '.jp-connection__disconnect-dialog__link' );
+
+		it( 'the connection link calls the trackModalClick with learn_about', () => {
+			links.at( learnAboutLink ).simulate( 'click', { preventDefault: () => undefined } );
```
Matching links by index seems like it could be fragile if the link order or content changes in the future.
I wonder if we'd be able to find the components `<jpConnectionInfoLink>` or `<jpSupportLink>`. I haven't tested that out, but that may be one way to specifically identify the links to test.

jim-coffey replied:
I agree they're a little brittle as written - but the short answer to your question is no.

`wrapper.debug()` shows the rendered html and in a quick test it wasn't possible to find these components.

jim-coffey replied:
I've changed to a loop which removes the brittleness,  while ensuring all events are tracked.

jboland88 commented on projects/js-packages/connection/components/disconnect-dialog/index.jsx:
```diff
@@ -200,6 +200,11 @@ const DisconnectDialog = props => {
 		e => {
 			e && e.preventDefault();
 
+			jetpackAnalytics.tracks.recordEvent(
+				'jetpack_disconnect_dialog_click_disconnect',
+				defaultTracksArgs
+			);
+
```
What would you think about having the disconnect button using a local event callback method inside the `<StepDisconnect />` component? That could allow a click on the disconnect button to call both `onDisconnect()` as well as `trackModalClick()`.
This could keep the analytics code in the `trackModalClick` function and also allow for clicking on the Disconnect button to be added to the test suite.

jim-coffey replied:
Good idea - done.

jboland88 commented on projects/js-packages/connection/components/disconnect-dialog/steps/step-disconnect.jsx:
```diff
@@ -30,8 +30,18 @@ const StepDisconnect = props => {
+		trackModalClick,
+	] );
+	const trackSupportClick = useCallback( () => trackModalClick( 'support' ), [ trackModalClick ] );
+	const trackStayConnectedClick = useCallback( () => {
```
Bit of a nitpick - thoughts on naming this callback to something like `handleStayConnectedClick` since it both captures an event and closes the modal?

jboland88 reviewed (changes requested):
Looks good overall, added a few comments inline 

Some small notes on the test plan:
> You can see if tracking events are being triggered by looking for the tracking pixels in the Network and checking query parameters

Some more instruction on this could be helpful on this point, is there something to filter by in the Network panel that would show the requests that we are looking for?

A screenshot of this modal from a testing site would be a nice add to the test plan to show which links are being tested. I think the steps are descriptive for what to click/ observe, but a visual would help confirm the tester is looking at the right thing.

Since JS unit tests are being added/ modified, we'll want to be sure to include an instruction to confirm the tests are passing as part of the test plan. IIRC, to run tests for this package, you can run `pnpm test` from within the package directory.

jim-coffey committed:
changes based on review comments

- rename stay connected handler
- move all click tracking code to step-disconnect

jim-coffey commented:
> It all works great. Nice work, Jim. I left a few inline questions and comments.

Hey @travisw - I'm probably missing something - but I can't see any comments from you 🤔

jim-coffey committed:
Update based on comments

- fix tests to loop and be less brittle
- stop constructing event names programmatically
- some formatting

jim-coffey committed:
Merge branch 'master' into add/disconnect-modal-click-events

jim-coffey commented on projects/js-packages/connection/components/disconnect-dialog/steps/step-disconnect.jsx:
```diff
@@ -30,8 +30,25 @@ const StepDisconnect = props => {
+			onDisconnect( e );
+		},
+		[ trackModalClick, onDisconnect ]
+	);
```
I've updated these to be a bit more consistent.

Any inconsistency now was due to Prettier re-formatting :smile:

jim-coffey commented on projects/js-packages/connection/components/disconnect-dialog/index.jsx:
```diff
@@ -220,6 +220,16 @@ const DisconnectDialog = props => {
+	const trackModalClick = useCallback(
+		target => {
+			jetpackAnalytics.tracks.recordEvent(
+				`jetpack_disconnect_dialog_click_${ target }`,
```
Interesting point - I've updated these to call the full event names explicitly 👍

jim-coffey commented on projects/js-packages/connection/components/disconnect-dialog/steps/step-disconnect.jsx:
```diff
@@ -30,8 +30,25 @@ const StepDisconnect = props => {
+		trackModalClick,
 	} = props;
 
+	const trackLearnClick = useCallback( () => trackModalClick( 'learn_about' ), [
```
For this one, we would still have to pass through the default tracking args from the parent component (or recreate which would be messy duplication).

So:
  - we have to pass something, so the function is nice and simple
  - we don't have to re-import the analytics functions in the child component
  - the child component calls are even simpler and don't have to worry about default analytics args
  - it makes the testing a bit simpler as we can spy on the function

jim-coffey committed:
version fixup script

jboland88 commented on projects/js-packages/connection/components/disconnect-dialog/test/step-disconnect.jsx:
```diff
@@ -35,25 +40,66 @@ describe( 'StepDisconnect', () => {
+		];
+		const wrapper = shallow( <StepDisconnect { ...testProps } /> );
+		const links = wrapper.find( '.jp-connection__disconnect-dialog__link' );
+		let trackedTotal = 0;
+
+		links.forEach( ( link, i ) => {
+			link.simulate( 'click', { preventDefault: () => undefined } );
+
+			const clickTrackValue = testProps.trackModalClick.getCall( i ).args[ 0 ];
+
+			it( `should track ${ clickTrackValue }`, async () => {
+				const valueIdx = trackEvents.indexOf( clickTrackValue );
+
+				expect( clickTrackValue ).to.equal( trackEvents[ valueIdx ] );
+				trackedTotal++;
+			} );
+		} );
+
+		it( `should track all ${ trackEvents.length } expected events`, () => {
+			expect( trackedTotal ).to.equal( trackEvents.length );
+		} );
```
```suggestion
		
		it( 'Calls all expected help link events', () => {
			links.forEach( link => {
				link.simulate( 'click', { preventDefault: () => undefined } );
			} );

			trackEvents.forEach( event_name => {
				expect( testProps.trackModalClick.calledOnceWith( event_name ) );
			} );
		} );
```

This test may not need to be concerned about the order of the events/ callbacks? I think displaying the event names as part of the test name is a nice add, but may be making the test too complex.

jim-coffey replied:
I like the simplicity of your test 👍 

The order shouldn't matter in my version either. I also wanted to try and catch if people added links without tracking, or tracking links without testing... so I'll leave it as it is.

jboland88 reviewed (dismissed):
I think this looks good and tests well.
Added one suggestion around the unit tests, but not considering that a blocker.

jim-coffey committed:
Merge branch 'master' into add/disconnect-modal-click-events

jeherve reviewed (dismissed):
This tests well for me. 🚢 
The js tests are failing at the moment, but it may be a false alert. I've relaunched the tests just now.

jim-coffey committed:
Merge branch 'master' into add/disconnect-modal-click-events

jim-coffey committed:
[not verified] Merge branch 'master' into add/disconnect-modal-click-events
